src/main/scala/org/infinispan/spark/rdd/FilteredInfinispanRDD.scala

```diff
@@ -0,0 +1,88 @@
+package org.infinispan.spark.rdd
+
+import java.net.InetSocketAddress
+
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.rdd.RDD
+import org.apache.spark.{Partition, TaskContext}
+import org.infinispan.client.hotrod.RemoteCacheManager
+import org.infinispan.client.hotrod.filter.Filters
+import org.infinispan.client.hotrod.marshall.ProtoStreamMarshaller
+import org.infinispan.protostream.annotations.ProtoSchemaBuilder
+import org.infinispan.protostream.{BaseMarshaller, FileDescriptorSource}
+import org.infinispan.query.dsl.Query
+import org.infinispan.spark._
+
+import scala.collection.JavaConversions._
+
+/**
+ * Infinispan RDD based on pre-filtered data in a cache, either by a Query or a filter previously deployed in the server
+ *
+ * @author gustavonalle
+ */
+class FilteredInfinispanRDD[K, V, R](parent: InfinispanRDD[K, V],
+                                     filterQuery: Option[FilterQuery],
+                                     filterFactory: Option[String],
+                                     filterParams: AnyRef*)
+        extends RDD[(K, R)](parent.sc, Nil) {
+
+   def createBuilder(address: InetSocketAddress) = {
+      val builder = parent.createBuilder(address)
+      if (filterQuery.isDefined) builder.marshaller(new ProtoStreamMarshaller)
+      builder
+   }
+
+   override def count() = {
+      if (filterFactory.isDefined) super.count()
+      else
+         filterQuery.fold(parent.remoteCache.size)(_.queryObj.getResultSize)
+   }
+
+   @DeveloperApi
+   override def compute(split: Partition, context: TaskContext): Iterator[(K, R)] = {
+      val (factory, params) = filterQuery.fold((filterFactory.orNull, filterParams))(q => (Filters.ITERATION_QUERY_FILTER_CONVERTER_FACTORY_NAME, Array(q.query)))
+      parent.compute(split, context, address => {
+         val cm = new RemoteCacheManager(createBuilder(address).build())
```

Comment:
What's the lifecycle of this `RemoteCacheManager` like? IOW, are you making sure it gets closed properly? `RemoteCacheManager` holds resources like connections pools, thread pools, so you don't want to leak those.